machineconfiguration/v1alpha1/types_machineosconfig.go

```diff
@@ -0,0 +1,233 @@
+	// +kubebuilder:validation:Required
+	RenderedImagePullspec string `json:"renderedImagePullspec"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// +patchMergeKey=containerFileArch
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=containerFileArch
+	// +optional
+	// +kubebuilder:validation:MinItems=0
+	// +kubebuilder:validation:MaxItems=2
+	// +optional
+	Containerfile []MachineOSContainerfile `json:"containerFile" patchStrategy:"merge" patchMergeKey:"containerFileArch"`
+}
+
+// MachineOSContainerfile contains all custom content the user wants built into the image
+type MachineOSContainerfile struct {
+	// containerFileArch describes the architecture this containerfile is to be built for
+	// this arch is optional. If the user does not specify an architecture, it is assumed
+	// that the content can be applied to all architectures, or in a single arch cluster: the only architecture.
+	// +kubebuilder:validation:Enum:=arm64;amd64;ppc64le;s390x;aarch64;x86_64;noArch
+	// +kubebuilder:default:=noArch
+	// +optional
+	ContainerFileArch ContainerFileArch `json:"containerFileArch"`
+	// content is the custom content to be built
+	// +kubebuilder:validation:Required
+	Content string `json:"content"`
+}
+
+type ContainerFileArch string
+
+const (
+	// describes the arm64 architecture
+	Arm64 ContainerFileArch = "arm64"
+	// describes the amd64 architecture
+	Amd64 ContainerFileArch = "amd64"
+	// describes the ppc64le architecture
+	Ppc ContainerFileArch = "ppc64le"
+	// describes the s390x architecture
+	S390 ContainerFileArch = "s390x"
+	// describes the aarch64 architecture
+	Aarch64 ContainerFileArch = "aarch64"
+	// describes the x86_64 architecture
+	x86_64 ContainerFileArch = "x86_64"
+	// describes a containerfile that can be applied to any arch
+	noArch ContainerFileArch = "noArch"
+)
+
+// Refers to the name of a MachineConfigPool (e.g., "worker", "infra", etc.):
+type MachineConfigPoolReference struct {
+	// name of the MachineConfigPool object.
+	// +kubebuilder:validation:MaxLength:=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of an image registry push/pull secret needed in the build process.
+type ImageSecretObjectReference struct {
+	// name is the name of the secret used to push or pull this MachineOSConfig object.
```

Comment:
which  namespace is this in?  having a hardcoded requirement makes sense.  We've generally used `openshift-config` for ones we don't own.

Reply:
the secret itself? I believe we use two namespaces. one lives in MCO namespace, the other is created without a namespace which I am assuming defaults to some sane namespace (openshift-config). For the MCO one, we use the pull secret found in openshift-config and pull out data from it, then creating a copy in MCO...

I am open to changing this and putting the all in openshift-config, if so: how would I namespace this?